src/main/java/org/opensearch/knn/quantization/quantizer/Quantizer.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.quantization.quantizer;
+
+import org.opensearch.knn.quantization.models.quantizationOutput.QuantizationOutput;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+import org.opensearch.knn.quantization.models.requests.TrainingRequest;
+
+/**
+ * The Quantizer interface defines the methods required for training and quantizing vectors
+ * in the context of K-Nearest Neighbors (KNN) and similar machine learning tasks.
+ * It supports training to determine quantization parameters and quantizing data vectors
+ * based on these parameters.
+ *
+ * @param <T> The type of the vector or data to be quantized.
+ * @param <R> The type of the quantized output, typically a compressed or encoded representation.
+ */
+public interface Quantizer<T, R> {
```

Comment:
Based on method signature, this seems to a bifunction, is it worth re-considering?

```
public interface Quantizer<T, U extends QuantizationState, R> 
```

Reply:
Quantization State is something Type Independent , It wil be overKill if I define that as U , right now there are no strong use cases requiring different types of QuantizationState or QuantizationOutput, the added complexity might not be necessary. Can revisit in future if needed.

Reply:
```public interface QuantizationState extends Externalizable {
    /**
     * Returns the quantization parameters associated with this state.
     *
     * @return the quantization parameters.
     */
    QuantizationParams getQuantizationParams();

    /**
     * Serializes the quantization state to a byte array.
     *
     * @return a byte array representing the serialized state.
     * @throws IOException if an I/O error occurs during serialization.
     */
    byte[] toByteArray() throws IOException;
}
```

This is quantization State , It is not any Type specific, I really don't want to abuse generics and follow KISS.